Splitview for Reader

frosty opened the pull request:
Implements `UISplitViewController` for the Reader section of the app.

The left pane always shows the list of streams and the right pane shows stream contents, post details, comments, etc. This should behave the same as Calypso's Reader.

To test:

* Check everything works as you'd expect, including with multitasking and on iPhone.
* When you add a new tag, it should become selected in the list and get scrolled into view. If you're on an iPad and the splitview is expanded then you should be shown the content of the tag. If the splitview is collapsed or you're on iPhone, then it'll just be highlighted and then deselected.
* Check things get reset when signing out of the app and into a different account.
* State restoration should work!

This PR doesn't include fullscreen for reader post details – that's coming in a separate PR (next week).

Needs review: @kurzee 

@aerych Would you mind double checking that things look as you'd expect too?

aerych commented:
Hola @frosty 
A split screen reader is so cool to see :D 
After a quick test drive there were just two things I noticed. 
1. After logging into a wpcom account for the first time the Followed topic is the one that's selected/shown. The default topic should be Discover.
2. After tapping Add a Tag, the modal animates up, then there is a momentary stutter before the keyboard appears. 

Tested on a device: iPad Pro 12.9"

kurzee commented:
Very nice @frosty!

Quick note, it would be really nice if we could hang on to cell selection when popping back to the `ReaderMenuViewController`. I personally expect this as a default iOS behavior, but it really helps to keep track of where a user was at when pushing and popping.

kurzee commented:
Not sure if this would be affected from this PR or not, but I can't get a Reader detail post to restore when testing. It does restore to the root stream view controller, but not to a selected detail view controller.

kurzee commented:
@frosty I think is good to go, I can't seem to break it any other way. Just the issues mentioned above if you want to address now or later.

frosty commented:
@aerych Thanks for testing!

* I'll make Discover the first topic – thanks for pointing that out.
* The stutter before the keyboard appears seems to be because `SettingsTextViewController` sets the text field as first responder in `viewDidAppear`. Changing it to `viewWillAppear` fixes the issue – not sure whether to go ahead and do this as `SettingsTextViewController` is presented from quite a few places where it seems to be working okay. I tested it in a few places and it seemed okay, but would be good to get a second opinion.

frosty commented:
@kurzee When you say about hanging on to cell selection – you mean if you're on a phone and you pop back up to the reader menu, you'd like to highlight the cell of whichever section you came from? We could do this, but I don't feel like it's normal iOS behaviour... I think personally I'd prefer not to, but let me know if you feel strongly :D 

Looking into restoration.

kurzee commented:
@frosty normal iOS behavior in that cell selections would stay selected until coming back to the view. Which is where `clearsSelectionOnViewWillAppear` generally takes care of it as long as the cell wasn't explicitly deselected in `selectedIndexPathForRow:`.

I'd say I'd love to have it for any standard looking tableViews, such as this. Just looking at Settings, Mail, Messages, 1Password, etc. However, it's not _always_ used, so not that big of a deal if it's a major pain or hack to do.

frosty commented:
@kurzee Looks like the state restoration problem is a pre-existing issue and only affects some posts. Opened a separate issue: https://github.com/wordpress-mobile/WordPress-iOS/issues/6189

kurzee commented:
@frosty other than the cell selection thing, which could be done separately, I think this is good to go! :shipit:

frosty commented:
@kurzee FINALLY revisiting this, and it'd be great if you could do me another review. I've fixed up everything (I think) apart from the cell selection thing – and I think that might be due to the tableview reloading its content in a couple of different places, so I'd rather look at that separately. Addressed in the last updates:

* State restoration should be more reliable (although some Discover posts won't restore – https://github.com/wordpress-mobile/WordPress-iOS/issues/6189). 
* Reader now always displays the detail pane when the split view collapses. Other sections should only show the detail pane if you interacted with it (pushed / popped) or selected a new item in the left pane.
* Reader now shows Discover by default instead of Followed Sites.

If you could check things look right on phones (including Plus), iPads, multitasking... that'd be awesome. Almost there!

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift:
```diff
@@ -229,18 +229,22 @@ import WordPressShared
+        showDetailViewController(viewControllerForTopic(topic), sender: self)
     }
 
+    private func viewControllerForTopic(topic: ReaderAbstractTopic) -> UIViewController {
```
Would we want to type the return here with `ReaderStreamViewController`? The method does seem to infer some kind of generic type though, so no biggie.

kurzee commented on WordPress/Classes/ViewRelated/Reader/ReaderMenuViewController.swift:
```diff
@@ -229,18 +229,22 @@ import WordPressShared
+        showDetailViewController(viewControllerForSearch(), sender: self)
     }
 
+    private func viewControllerForSearch() -> UIViewController {
```
Same for this one, maybe type with `ReaderSearchViewController`?

kurzee commented on WordPress/Classes/ViewRelated/System/WPSplitViewController.swift:
```diff
@@ -362,9 +362,19 @@ extension WPSplitViewController: UISplitViewControllerDelegate {
         if let primaryViewController = primaryViewController as? UINavigationController {
-            if let secondaryViewController = secondaryViewController as? UINavigationController
-                where detailNavigationStackHasBeenModified || collapseMode == .AlwaysKeepDetail {
+            if let secondaryViewController = secondaryViewController as? UINavigationController{
```
Maybe just a space between `troller {` on `UINavigationController{`. 🤓

kurzee commented:
@frosty this is working excellently!!

I tested:

- iPad multitasking in portrait and landscape
- iPhone Plus orientation changes
- State restoration on iPhone and iPad
- State restoration of nested detail post view controllers
- State restoration with orientation changes
- State restoration when restoring via multitasking
- Log out/log ins
- Fresh installs

This is good to go after the minor tweaks mentioned above (or without). Nice work.

:shipit: when ready, or ping me if you push more tweaks!

Note: milestone is currently `6.8` and targeting `develop`.
